mne/beamformer/_compute_beamformer.py

```diff
@@ -253,96 +228,94 @@ def _compute_beamformer(G, Cm, reg, n_orient, weight_norm, pick_ori,
-            signs[signs == 0] = 1.
-            max_power_ori *= signs[:, np.newaxis]
-            # all three entries get the same value from this operation
-            Wk[:] = np.sum(max_power_ori[:, :, np.newaxis] * Wk, axis=1,
-                           keepdims=True)
-    W = Wk.reshape(n_sources * n_orient, n_channels)
-    del Gk, Wk, sk
-
-    if pick_ori == 'normal':
-        W = W[2::3]
-    elif pick_ori == 'max-power':
-        W = W[0::3]
-
-    # Re-scale the filter weights according to the selected weight
-    # normalization scheme
-    if weight_norm in ['unit-noise-gain', 'nai']:
-        if pick_ori in [None, 'vector'] and n_orient > 1:
-            # Rescale each set of 3 filters
-            W = W.reshape(-1, 3, W.shape[1])
-            noise_norm = np.linalg.norm(W, axis=(1, 2), keepdims=True)
+    def _compute_bf_terms(Gk, Cm_inv):
+        bf_numer = np.matmul(Gk.transpose(0, 2, 1), Cm_inv[np.newaxis])
+        bf_denom = np.matmul(bf_numer, Gk)
+        return bf_numer, bf_denom
+
+    #
+    # 2. Reorient lead field in direction of max power or normal
+    #
+    if pick_ori == 'max-power':
+        assert n_orient == 3
+        _, bf_denom = _compute_bf_terms(Gk, Cm_inv)
+        if weight_norm is None:
+            ori_numer = np.eye(n_orient)[np.newaxis]
+            ori_denom = bf_denom
         else:
-            # Rescale each filter separately
-            noise_norm = np.linalg.norm(W, axis=1, keepdims=True)
+            assert weight_norm in ['unit-noise-gain', 'nai']
+            # compute power, cf Sekihara & Nagarajan 2008, eq. 4.47
+            ori_numer = bf_denom
+            ori_denom = np.matmul(
+                np.matmul(Gk.transpose(0, 2, 1),
+                          Cm_inv.dot(Cm_inv)[np.newaxis]), Gk)
+        ori_denom_inv = _sym_inv_sm(ori_denom, reduce_rank, inversion, sk)
+        ori_pick = np.matmul(ori_denom_inv, ori_numer)
+        assert ori_pick.shape == (n_sources, n_orient, n_orient)
+
+        # pick eigenvector that corresponds to maximum eigenvalue:
+        eig_vals, eig_vecs = np.linalg.eig(ori_pick.real)  # not Hermitian!
+        # sort eigenvectors by eigenvalues for picking:
+        order = np.argsort(np.abs(eig_vals), axis=-1)
+        # eig_vals = np.take_along_axis(eig_vals, order, axis=-1)
+        max_power_ori = np.take_along_axis(
+            eig_vecs, order[:, np.newaxis, -1:], axis=-1)[:, :, 0]
+        assert max_power_ori.shape == (n_sources, n_orient)
+
+        # set the (otherwise arbitrary) sign to match the normal
+        signs = np.sign(np.sum(max_power_ori * nn, axis=1, keepdims=True))
+        max_power_ori *= signs
+
+        # Compute the lead field for the optimal orientation,
+        # and adjust numer/denom
+        Gk = np.matmul(Gk, max_power_ori[..., np.newaxis])
+        n_orient = 1
+    else:
+        max_power_ori = None
+        if pick_ori == 'normal':
+            Gk = Gk[..., 2:3]
+            n_orient = 1
```

Comment:
Moved `pick_ori='normal'` up here, now it uses gain matrix subselection just like `pick_ori='max-power'`, and from then on the formulas proceed like scalar beamformers.

Again, don't know if it's correct, but it's more consistent...

Reply:
I just kept `pick_ori='normal'` along where it used to be - I am not very familiar with it as I always compute beamformers in volume source space so this never really comes up. It is not an option that is seen very often in the wild with beamformer either I think - not saying it's not useful, but not seen often. If we are not sure it is correct, is it worth keeping? CC @wmvanvliet

Reply:
Actually come to think of it I think this is a reasonable thing to do. You should get the same result from this as you would from passing a fixed-orientation forward. If you want the normal component from a vector free-orientation solution, you should use `stc_free.normal()` and this will differ from the fixed-orientation-forward result in the minimum-norm case, too. So I'll mark this one as resolved